virtio-devices/src/transport/pci_device.rs

```diff
@@ -939,39 +1006,43 @@ impl PciDevice for VirtioPciDevice {
             .set_address(virtio_pci_bar_addr.raw_value())
             .set_size(CAPABILITY_BAR_SIZE)
             .set_region_type(region_type);
-        self.configuration.add_pci_bar(&bar).map_err(|e| {
-            PciDeviceError::IoRegistrationFailed(virtio_pci_bar_addr.raw_value(), e)
-        })?;
+        if !restoring {
+            self.configuration.add_pci_bar(&bar).map_err(|e| {
+                PciDeviceError::IoRegistrationFailed(virtio_pci_bar_addr.raw_value(), e)
+            })?;
 
-        bars.push(bar);
+            // Once the BARs are allocated, the capabilities can be added to the PCI configuration.
+            self.add_pci_capabilities(VIRTIO_COMMON_BAR_INDEX as u8)?;
+        }
```

Comment:
Can you explain why this is only needed for non-restoring? May worth adding a comment.